File: android_client/test/radio_editions_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:android_client/radio_editions.dart';
import 'package:android_client/src/rust/models.dart';

void main() {
  test('complete program replaces old category tracks once audio is ready', () {
    const old=Item(id:'old',title:'old',audioUrl:'/audio/old.mp3',durationSec:60,tags:'["radio:date:2026-09-22","radio:edition:morning"]');
    const full=Item(id:'full',title:'full',audioUrl:'/audio/full.mp3',durationSec:600,tags:'["radio:date:2026-09-22","radio:edition:morning","radio:program","radio:sections:11"]');
    final group=buildRadioEditions([old,full]).single;
    expect(group.playbackIds,['full']);expect(group.totalDurationSec,600);expect(programSectionCount(full),11);
  });
  test(
    'Logical morning survives midnight TTS and keeps waiting audio out of totals',
    () {
      const ready = Item(
        id: 'a',
        title: 'Morning',
        publishTime: 1790006400,
        audioUrl: '/audio/a.mp3',
        durationSec: 61,
        tags: '["radio:date:2026-09-21","radio:edition:morning"]',
      );
      const waiting = Item(
        id: 'b',
        title: 'Waiting',
        publishTime: 1790006400,
        durationSec: 999,
        tags: '["radio:date:2026-09-21","radio:edition:morning"]',
      );
      const evening = Item(
        id: 'c',
        title: 'Evening',
        publishTime: 1790006400,
        audioUrl: '/audio/c.mp3',
        durationSec: 90,
        tags: '["radio:date:2026-09-21","radio:edition:evening"]',
      );
      final groups = buildRadioEditions([ready, waiting, evening]);
      expect(groups.length, 2);
      final morning = groups.singleWhere((g) => g.key.endsWith('morning'));
      expect(morning.title, contains('2026年9月21日 星期一'));
      expect(morning.playbackIds, ['a']);
      expect(morning.totalDurationSec, 61);
      expect(morning.items.length, 2);
    },
  );
  test(
    'Legacy timestamps follow Shanghai date, independent of device timezone',
    () {
      final item = Item(
        id: 'a',
        title: 'Legacy',
        publishTime:
            DateTime.parse('2026-09-20T23:30:00Z').millisecondsSinceEpoch ~/
            1000,
      );
      expect(buildRadioEditions([item]).single.key, '2026-09-21:morning');
    },
  );
}

File: scripts/radio-editions.test.ts
import { strict as assert } from 'node:assert';
import { test } from 'node:test';
import { buildRadioEditions } from '../frontend/src/radio-editions.ts';
import type { Item } from '../frontend/src/types.ts';
const make = (id: string, edition: string, audio: boolean, duration: number): Item => ({id, title:id, summary:null, original_url:null, cover_image_url:null, publish_time:1790006400, created_at:1790006400, tags:JSON.stringify(['radio:date:2026-09-21',`radio:edition:${edition}`]), audio_url:audio?`/audio/${id}.mp3`:null, duration_sec:duration});
test('logical editions survive late generation and count only playable duration',()=>{
  const groups=buildRadioEditions([make('m1','morning',true,61),make('m2','morning',false,900),make('e','evening',true,125)]);
  assert.equal(groups.length,2);
  const morning=groups.find(g=>g.key.endsWith('morning'))!;
  assert.match(morning.title,/2026年9月21日星期一/);
  assert.deepEqual(morning.playbackIds,['m1']);
  assert.equal(morning.totalDurationSec,61);
  const updated=buildRadioEditions([make('m1','morning',true,61),make('m2','morning',true,90)])[0];
  assert.equal(updated.playableCount,2); assert.equal(updated.totalDurationSec,151);
});
test('legacy timestamp uses Shanghai regardless of device timezone',()=>{
  const item={...make('x','morning',true,60),tags:null,publish_time:Date.parse('2026-09-20T23:30:00Z')/1000};
  assert.equal(buildRadioEditions([item])[0].key,'2026-09-21:morning');
});

test('a ready complete program replaces legacy tracks without double duration',()=>{
  const old=make('old','morning',true,61);
  const program={...make('program','morning',true,600),tags:JSON.stringify(['radio:date:2026-09-21','radio:edition:morning','radio:program','radio:sections:11'])};
  const group=buildRadioEditions([old,program])[0];
  assert.deepEqual(group.playbackIds,['program']);assert.equal(group.totalDurationSec,600);
  assert.deepEqual(buildRadioEditions([old,{...program,audio_url:null}])[0].playbackIds,['old']);
});
